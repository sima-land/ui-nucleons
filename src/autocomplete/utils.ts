import type { SyntheticEvent } from 'react';

/**
 * Наивная реализация синтетического события.
 * Не рекомендуется к использованию за пределами библиотеки.
 */
export class NaiveSyntheticEvent<T extends Element = Element> implements SyntheticEvent<T> {
  nativeEvent: Event;
  currentTarget: EventTarget & T;
  target: EventTarget & T;
  bubbles: boolean;
  cancelable: boolean;
  defaultPrevented: boolean;
  eventPhase: number;
  isTrusted: boolean;
  timeStamp: number;
  type: string;

  /** @inheritdoc */
  constructor(nativeEvent: Event, target: T) {
    this.target = target;
    this.currentTarget = target;
    this.nativeEvent = nativeEvent;
    this.type = nativeEvent.type;
    this.timeStamp = nativeEvent.timeStamp;
    this.bubbles = nativeEvent.bubbles;
    this.cancelable = nativeEvent.cancelable;
    this.defaultPrevented = nativeEvent.defaultPrevented;
    this.eventPhase = nativeEvent.eventPhase;
    this.isTrusted = nativeEvent.isTrusted;
  }

  /** @inheritdoc */
  persist() {
    return void 0;
  }

  /** @inheritdoc */
  stopPropagation() {
    this.nativeEvent.stopPropagation();
  }

  /** @inheritdoc */
  isPropagationStopped() {
    return false;
  }

  /** @inheritdoc */
  preventDefault() {
    this.nativeEvent.preventDefault();
    this.defaultPrevented = this.nativeEvent.defaultPrevented;
  }

  /** @inheritdoc */
  isDefaultPrevented() {
    return this.defaultPrevented;
  }
}

/**
 * Вызывает переданную функцию только если переданное значение не является null или undefined.
 * @param value Значение.
 * @param callback Функция.
 * @return Результат функции или undefined.
 */
export function when<A, B>(value: A, callback: (value: NonNullable<A>) => B): B | undefined {
  return value !== null && value !== undefined ? callback(value) : undefined;
}
