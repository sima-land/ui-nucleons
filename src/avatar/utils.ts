/**
 * Формирует монограмму из переданной строки.
 * @param value Переданная строка.
 * @return Монограмма.
 */
export function getMonogram(value: any): string {
  if (typeof value === 'string') {
    return value
      .split(/\s+/g)
      .slice(0, 2)
      .map(s => s[0])
      .reverse()
      .join('')
      .toUpperCase();
  }

  return '';
}
