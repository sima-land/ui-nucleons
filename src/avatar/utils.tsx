import type { SVGAttributes } from 'react';
import type { AvatarProps } from './types';
import PersonSVG from '@sima-land/ui-quarks/icons/24x24/Stroked/Person';
import classNames from 'classnames';
import styles from './utils.module.scss';

const USER_AVATAR_COLOR_TOKENS: ReadonlyArray<string> = [
  '#eb8585',
  '#fda09b',
  '#f49bb1',
  '#ffd785',
  '#ffbb85',
  '#89d7c7',
  '#85e5ac',
  '#b5ef90',
  '#93c4ee',
  '#85caf5',
  '#85ddea',
  '#b3c1c7',
  '#dc9bd5',
  '#b4a8f0',
  '#c999e9',
];

/**
 * Возвращает свойства для аватара пользователя.
 * @param payload Данные пользователя.
 * @param customProps Дополнительные свойства аватара.
 * @return Свойства.
 */
export function getUserAvatarProps(
  {
    id,
    name,
    image,
  }: {
    id?: number;
    name?: string;
    image?: string;
  },
  customProps?: AvatarProps,
): AvatarProps {
  return {
    src: image,
    children: getNameInitials(name) || <PersonSVG {...getAvatarIconProps()} />,
    ...customProps,
    className: classNames(styles.user, customProps?.className),
    style: {
      '--avatar-color': id ? getUserAvatarColor(`${id}`) : undefined,
      ...customProps?.style,
    },
  };
}

/**
 * Возвращает цвет аватара текущего пользователя.
 * @param identity Срока, идентифицирующая пользователя.
 * @return Цвет.
 */
export function getUserAvatarColor(identity: string): string {
  let total = 0;

  for (let i = 0; i < identity.length; i++) {
    total += identity.charCodeAt(i);
  }

  const list = USER_AVATAR_COLOR_TOKENS;
  const hash = total % list.length;

  return list[hash];
}

/**
 * Формирует инициалы имени.
 * @param value Строка.
 * @return Инициалы.
 */
export function getNameInitials(value: any): string {
  if (typeof value === 'string') {
    return (value[0] ?? '').toUpperCase();
  }

  return '';
}

/**
 * Возвращает свойства для иконки внутри аватара.
 * @return Свойства.
 */
export function getAvatarIconProps(): SVGAttributes<SVGSVGElement> {
  return {
    className: styles.icon,
  };
}
