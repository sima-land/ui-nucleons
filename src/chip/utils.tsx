import type { ChipProps } from './types';
import { ChipIconButton } from './chip';
import CrossSVG from '@sima-land/ui-quarks/icons/16x16/Stroked/Cross';

/**
 * Возвращает свойства для чипа, добавляющие ему кнопку удаления.
 * @param options Опции.
 * @return Свойства.
 */
export function getDeletableChipProps({ onDelete }: { onDelete?: VoidFunction } = {}): ChipProps {
  return {
    padding: 'start',
    adornmentGutter: 'unset',
    endAdornment: (
      <ChipIconButton type='button' onClick={onDelete}>
        <CrossSVG fill='currentColor' />
      </ChipIconButton>
    ),
  };
}
