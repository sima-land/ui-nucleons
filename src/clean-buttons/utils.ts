import type { CleanButtonProps } from './types';
import { createContext } from 'react';

/**
 * Контекст установки размера группы прозрачных кнопок CleanGroup.
 * Размер заданный непосредственно компоненту CleanGroup имеет приоритет над контекстом.
 */
export const CleanButtonContext = createContext<Pick<CleanButtonProps, 'size' | 'className'>>({});
