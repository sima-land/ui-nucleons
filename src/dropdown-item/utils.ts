import { isValidElement } from 'react';
import { DropdownItem } from './dropdown-item';
import { DropdownItemElement } from './types';

/** Утилиты для работы с DropdownItem. */
export class DropdownItemUtils {
  /**
   * Определяет, является ли переданное значение элементом DropdownItem.
   * @param value Значение.
   * @return Является ли переданное значение элементом DropdownItem.
   */
  static is(value: Parameters<typeof isValidElement>[0]): value is DropdownItemElement {
    return isValidElement(value) && value.type === DropdownItem;
  }

  /**
   * Возвращает значение элемента меню.
   * @param menuItem Значение.
   * @return Значение элемента меню.
   */
  static getValue(menuItem: DropdownItemElement) {
    const { value, children } = menuItem.props;

    // нативный select преобразует содержимое option к строке если не указан value, делаем также
    return typeof value === 'string' ? value : String(children);
  }

  /**
   * Получив опции меню вернет LinkedList и текущую позицию в нем.
   * @param menuItems Опции.
   * @param currentIndex Индекс текущей выбранной опции.
   * @return Текущая позиция в списке и список.
   */
  static asLoopedIterator(
    menuItems: DropdownItemElement[],
    currentIndex: number,
  ): LoopedIterator<number> {
    let currentPosition = -1;

    const availableIndices: Array<number> = menuItems.reduce<number[]>((acc, item, index) => {
      if (index === currentIndex) {
        currentPosition = acc.length;
      }

      if (!item.props.disabled) {
        acc.push(index);
      }

      return acc;
    }, []);

    return new LoopedIterator<number>(availableIndices, currentPosition);
  }
}

/** Список с возможностью перемещаться вперед назад с закольцовыванием. */
export class LoopedIterator<T = unknown> {
  private source: T[];
  private position: number;

  /**
   * Конструктор.
   * @param source Источник данных.
   * @param startPosition Начальная позиция.
   */
  constructor(source: T[], startPosition = 0) {
    this.source = source;
    this.position = startPosition;
  }

  /**
   * Возвращает следующий элемент списка.
   * @return Элемент списка или undefined.
   */
  next(): T | undefined {
    if (this.source.length === 0) {
      return undefined;
    }

    let nextIndex = this.position + 1;

    if (nextIndex >= this.source.length) {
      nextIndex = 0;
    }

    this.position = nextIndex;

    return this.source[nextIndex];
  }

  /**
   * Возвращает предыдущий элемент списка.
   * @return Элемент списка или undefined.
   */
  prev(): T | undefined {
    if (this.source.length === 0) {
      return undefined;
    }

    let nextIndex = this.position - 1;

    if (nextIndex < 0) {
      nextIndex = this.source.length - 1;
    }

    this.position = nextIndex;

    return this.source[nextIndex];
  }
}
