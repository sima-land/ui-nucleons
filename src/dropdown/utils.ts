import { useContext, useEffect, useMemo, useState } from 'react';
import { UseFloatingOptions, flip, autoUpdate, shift, UseFloatingReturn } from '@floating-ui/react';
import { useLayer } from '../helpers/layer';
import { ResizeObserverContext } from '../context';

/**
 * Возвращает конфигурацию для `useFloating` по дизайн-гайдам.
 * @return Конфигурация для `useFloating` по дизайн-гайдам.
 */
export function dropdownFloatingConfig(): Partial<UseFloatingOptions> {
  return {
    whileElementsMounted: autoUpdate,
    placement: 'bottom-start',
    middleware: [
      // @todo в будущем с дизайнерами можем решить использовать autoPlacement
      // но есть сложности: https://github.com/floating-ui/floating-ui/discussions/2155
      flip({
        padding: 16,
      }),
      shift({
        padding: 16,
      }),
    ],
  };
}

/**
 * Возвращает стили для элемента с учетом "слоя".
 * @param floating Результат вызова useFloating.
 * @return Стили для элемента с учетом "слоя".
 */
export function useDropdownFloatingStyle({
  strategy,
  x,
  y,
  refs,
}: Pick<UseFloatingReturn, 'strategy' | 'x' | 'y' | 'refs'>) {
  const layer = useLayer();

  const [openerWidth, setOpenerWidth] = useState<number | null>(
    () => refs.reference.current?.getBoundingClientRect().width ?? null,
  );

  const { createResizeObserver } = useContext(ResizeObserverContext);

  useEffect(() => {
    const opener = refs.reference.current;

    if (!(opener instanceof Element)) {
      return;
    }

    const observer = createResizeObserver(() => {
      setOpenerWidth(opener.getBoundingClientRect().width);
    });

    setOpenerWidth(opener.getBoundingClientRect().width);

    observer.observe(opener);

    return () => {
      observer.disconnect();
    };
  }, [refs.reference.current, createResizeObserver]);

  return useMemo(() => {
    const base = {
      position: strategy,
      top: y,
      left: x,
      zIndex: layer + 2,
    };

    if (typeof openerWidth === 'number') {
      return {
        ...base,

        minWidth: `${openerWidth}px`,

        // ВАЖНО: для возможности стилизации задаем переменную с шириной
        // @todo при необходимости добавить высоту и другие размеры или же наоборот отказаться от этой переменной
        '--opener-width': `${openerWidth}px`,
      };
    }

    return base;
  }, [strategy, x, y, layer, openerWidth]);
}
