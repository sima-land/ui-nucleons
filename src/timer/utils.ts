import type { TimeDistance } from './types';
import { formatDistanceToNowStrict } from 'date-fns/formatDistanceToNowStrict';
import { isAfter } from 'date-fns/isAfter';
import { isValid } from 'date-fns/isValid';
import { parseISO } from 'date-fns/parseISO';

const UNITS = ['day', 'hour', 'minute', 'second'] as const;

/**
 * Форматирует оставшееся время в виде "дни:часы:минуты:секунды".
 * @param distance Оставшееся время.
 * @return Отформатированное время.
 */
export function formatDistance({ days, hours, minutes, seconds }: TimeDistance): string {
  return [days, ...[hours % 24, minutes % 60, seconds % 60].map(s => `${s}`.padStart(2, '0'))].join(
    ':',
  );
}

/**
 * Возвращает объект с данными об оставшемся времени от текущего момента.
 * @param dateString Строка с датой в формате ISO.
 * @return Оставшееся время.
 */
export function getDistanceToNow(dateString: string): TimeDistance {
  const date = parseISO(dateString);

  let result = { days: 0, hours: 0, minutes: 0, seconds: 0 };

  if (isValid(date) && !isAfter(new Date(), date)) {
    const [days, hours, minutes, seconds] = UNITS.map(unit =>
      parseInt(
        formatDistanceToNowStrict(date, {
          unit,
          roundingMethod: 'floor',
        }).replace(/\D/g, ''),
      ),
    );

    result = { days, hours, minutes, seconds };
  }

  return result;
}
