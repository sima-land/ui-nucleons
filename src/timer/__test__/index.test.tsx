import { act, render } from '@testing-library/react';
import { getDistanceToNow } from '../utils';
import { Timer } from '..';

jest.useFakeTimers();
jest.spyOn(global, 'clearInterval');

jest.mock('../utils', () => {
  const original = jest.requireActual('../utils');

  return {
    ...original,
    __esModule: true,
    getDistanceToNow: jest.fn(() => ({ days: 1, hours: 2, minutes: 3, seconds: 4 })),
  };
});

beforeEach(() => {
  jest.clearAllTimers();
  jest.clearAllMocks();
});

describe('Timer', () => {
  it('should render properly without props', () => {
    const { container } = render(<Timer date='' />);

    expect(container).toMatchSnapshot();
  });

  it('should pass props', () => {
    const { container } = render(
      <Timer date='2030-03-08' timeout={500} format={({ days }) => `Осталось дней: ${days}`} />,
    );

    expect(container.textContent).toBe('Осталось дней: 1');
  });

  it('should change time by timeout', () => {
    render(<Timer date='2030-03-08' />);

    expect(getDistanceToNow).toHaveBeenCalledTimes(3);

    act(() => {
      jest.advanceTimersByTime(5000);
    });

    expect(getDistanceToNow).toHaveBeenCalledTimes(9);
  });

  it('should reset interval on unmount', () => {
    const { unmount } = render(<Timer date='30' />);

    expect(clearInterval).toHaveBeenCalledTimes(0);

    unmount();

    expect(clearInterval).toHaveBeenCalledTimes(1);
  });
});
