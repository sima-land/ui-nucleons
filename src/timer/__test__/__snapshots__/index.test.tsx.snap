// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer should render properly without props 1`] = `
<div>
  1:02:03:04
</div>
`;
